src/api/BundleDeploy/ParmValidator.ts

```diff
@@ -0,0 +1,237 @@
+* Eclipse Public License v2.0 which accompanies this distribution, and is available at
+* https://www.eclipse.org/legal/epl-v20.html
+*
+* SPDX-License-Identifier: EPL-2.0
+*
+* Copyright IBM Corp, 2019
+*
+*/
+
+"use strict";
+
+import { IHandlerParameters, Logger } from "@brightside/imperative";
+
+
+/**
+ * Class to represent a CICS Bundle manifest.
+ *
+ * @export
+ * @class Manifest
+ */
+export class ParmValidator {
+
+  public static validateDeploy(params: IHandlerParameters) {
+
+    // Validate the parms that are valid for Deploy
+    ParmValidator.validateName(params);
+    ParmValidator.validateBundledir(params);
+    ParmValidator.validateCicsDeployProfile(params);
+    ParmValidator.validateCicsplex(params);
+    ParmValidator.validateScope(params);
+    ParmValidator.validateCsdgroup(params);
+    ParmValidator.validateResgroup(params);
+    ParmValidator.validateTimeout(params);
+  }
+
+  public static validateUndeploy(params: IHandlerParameters) {
+
+    // Validate the parms that are valid for Undeploy
+    ParmValidator.validateName(params);
+    ParmValidator.validateCicsDeployProfile(params);
+    ParmValidator.validateCicsplex(params);
+    ParmValidator.validateScope(params);
+    ParmValidator.validateCsdgroup(params);
+    ParmValidator.validateResgroup(params);
+    ParmValidator.validateTimeout(params);
+  }
+
+  private static validateName(params: IHandlerParameters) {
+    // Name is mandatory
+    if (params.arguments.name === undefined) {
+      throw new Error("--name parameter is not set");
+    }
+
+    if (typeof params.arguments.name !== "string") {
+      throw new Error("--name parameter is not a string");
+    }
+
+    const MAX_LEN = 8;
+    if (params.arguments.name.length > MAX_LEN) {
+      throw new Error("--name parameter is too long");
+    }
+
+    if (params.arguments.name === "") {
+      throw new Error("--name parameter is empty");
+    }
+  }
+
+  private static validateBundledir(params: IHandlerParameters) {
+    // bundleDir is mandatory (but only for deploy. not undeploy)
+    if (params.arguments.bundledir === undefined) {
+      throw new Error("--bundledir parameter is not set");
+    }
+
+    if (typeof params.arguments.bundledir !== "string") {
+      throw new Error("--bundledir parameter is not a string");
+    }
+
+    const MAX_LEN = 255;
+    if (params.arguments.bundledir.length > MAX_LEN) {
+      throw new Error("--bundledir parameter is too long");
+    }
+
+    if (params.arguments.bundledir === "") {
+      throw new Error("--bundledir parameter is empty");
+    }
+  }
+
+  private static validateCicsDeployProfile(params: IHandlerParameters) {
+
+    // if missing, then CICSPlex and Scope must be set
+    if (params.arguments["cics-deploy-profile"] === undefined) {
+      if (params.arguments.cicsplex === undefined ||
+          params.arguments.scope === undefined) {
+          throw new Error("either --cics-deploy-profile or both --cicsplex and --scope must be set");
```

Comment:
I think it should be possible to set up a default profile and use that without explicitly specifying it - the base Zowe CLI commands work this way with the zosmf profile

Reply:
Correct, but... 

First issue is that these exceptions are arguably a waste of time anyway - Imperative wont even drive the Handler if it doesn't like the input. The validation is here for anyone calling the Handler/API directly.

The profile is configured to require both cicsplex and scope to be set; as a user it seems more friendly to me that they be told that either the profile or the two parms must be set; rather than simply saying that the two parms must be used. It integrates the profile more sensibly into what the user is told in failure scenarios. zowe doesn't do a particularly good job of integrating the profile into the user experience.

Reply:
Discussing with Paul, the handler will see the values from the profile as arguments. So this validation won't interfere with having a default profile